### PR TITLE
PROTOCOLS-117 LIST and NAMESPACE delegation

### PR DESCRIPTION
**Pull request by chibenwa**
Two **important** design choices are made in this PR:

  - Move mailboxName matching from mailbox to IMAP

This is done because we fully break the link between IMAP mailbox name and MailboxPath mailboxName by introducing virtual folders for delegation (goal of the PR BTW), thus matching on MailboxPath mailboxName **no longer makes sens**.

 - Protocol naming strategy should be orthogonal to mailboxPath naming. IE MailboxPath naming should be **absolute** and **unique for a single mailbox**.

The namespace definition has been moved from MailboxSession to ImapSession.

As you can see, these *two points greatly simplify James architecture*.

Also, I've been doing my boy scout with additional refactorings (aiming at removing as much coupling as possible, IE remove unneeded subclasses, etc...)

### CODE REVIEW

File: protocols/imap/src/test/java/org/apache/james/imap/processor/base/MailboxEventAnalyserTest.java [hunk 1]
```diff
@@ -19,734 +19,229 @@
...
-import org.apache.james.protocols.imap.DefaultNamespaceConfiguration;
+import org.junit.Before;
 import org.junit.Test;
 
 import com.google.common.collect.ImmutableList;
 
 public class MailboxEventAnalyserTest {
+    public static class SingleMessageResultIterator implements MessageResultIterator {
+        private final MessageResult messageResult;
+        boolean done;
```
Comment (aduprat) [inline protocols/imap/src/test/java/org/apache/james/imap/processor/base/MailboxEventAnalyserTest.java]: private

File: protocols/imap/src/test/java/org/apache/james/imap/processor/SetACLProcessorTest.java [hunk 2]
```diff
@@ -256,5 +179,28 @@ public void testRemoveRights() throws Exception {
...
+        verify(responder, times(1)).respond(argumentCaptor.capture());
+        verifyNoMoreInteractions(responder);
+
+        assertThat(argumentCaptor.getAllValues())
+            .hasSize(1);
+        assertThat(argumentCaptor.getAllValues().get(0))
+            .matches(StatusResponseTypeMatcher.OK_RESPONSE_MATCHER::matches);
+    }
 
 }
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply protocols/imap/src/test/java/org/apache/james/imap/processor/SetACLProcessorTest.java]: I don't understand... A quick look in the class imports shows there is no more dependency to JMock in that class.

The **OK_RESPONSE_MATCHER** is a hamcrest matcher wich is not depandant from JMOCK.

File: mailbox/api/src/main/java/org/apache/james/mailbox/PathDelimiter.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    public int lastIndex(String name) {
+        return name.lastIndexOf(charDelimiter);
+    }
+
+    public String appendDelimiter(String name) {
+        return name + charDelimiter;
+    }
+
+    public String removeTrailingSeparatorAtTheEnd(String name) {
+        if (name.length() == 0) {
```
Comment (aduprat) [inline mailbox/api/src/main/java/org/apache/james/mailbox/PathDelimiter.java]: `name.isEmpty`

File: mailbox/api/src/main/java/org/apache/james/mailbox/PathDelimiter.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            return name;
+        }
+        if (name.charAt(name.length() - 1) == charDelimiter) {
```
Comment (aduprat) [inline mailbox/api/src/main/java/org/apache/james/mailbox/PathDelimiter.java]: `name.endsWith`

File: mailbox/api/src/main/java/org/apache/james/mailbox/PathDelimiter.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+            return name.substring(0, name.length() - 1);
+        }
+        return name;
+    }
+
+    public String removeTrailingSeparatorAtTheBeginning(String name) {
+        if (name.length() == 0) {
```
Comment (aduprat) [inline mailbox/api/src/main/java/org/apache/james/mailbox/PathDelimiter.java]: `name.isEmpty`

File: mailbox/api/src/main/java/org/apache/james/mailbox/PathDelimiter.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+            return name;
+        }
+        if (name.charAt(0) == charDelimiter) {
```
Comment (aduprat) [inline mailbox/api/src/main/java/org/apache/james/mailbox/PathDelimiter.java]: `name.startsWith`

**Comment by mbaechler** [comment]
> Protocol naming strategy should be orthogonal to mailboxPath naming. IE MailboxPath naming should be absolute and unique for a single mailbox.

What is the value of MailboxPath when we already have MailboxId for that purpose ?

**Comment by chibenwa** [comment]
To answer you in short: mailboxes in "backend" has:
 -id : mailbox id
 -name: mailbox path.

The problem with protocol naming strategy is that we add "context" on a per user bases. Eg:

```
inbox.toto
Other users.mbeachler.toto
```

Are two valid IMAP level names. One for your perspective, one from mine. In order to have this context-sensitive naming, we should have a non ambiguous naming.

**Comment by mbaechler** [comment]
> Are two valid IMAP level names. One for your perspective, one from mine. In order to have this context-sensitive naming, we should have a non ambiguous naming.

Yes, that's imap MailboxName vs MailboxId. But MailboxPath looks really a duplication of MailboxId.

**Comment by chibenwa** [comment]
It starts being interesting other here!

Well, we (IMAP?) falls into the id vs name stuff. A name is still important, but should not be used to deal directly with a resource.

The fact that the names are different at the different layers justifies to have two name representation (string as mailbox name in IMAP, MailboxPath in backend akka mailbox).

However, what is not logic is to be able to perform ANY mailbox operation based on its mailboxPath. We should:
 - have a single name -> mailboxId function
 - handle rename, copy, move, deletes via mailboxId
 - fix aaaaall of the usages of these methods. 
 - deal with non mailbox id based implementation (maildir?)
 - completely replace MailboxPath by mailboxId in the event system. Feat #1085 ...
 
I would really be in favor of such a move. But I do believe that is not something we can handle "quickly in a pr".

(Side note: We might need strong typing in IMAP layer for mailbox name... I gonna be soooo happy to add yet another few thousand line diff ;-) )

File: protocols/imap/src/main/java/org/apache/james/imap/processor/ListProcessor.java [hunk 7]
```diff
@@ -89,156 +182,106 @@ protected ImapResponseMessage createResponse(boolean noInferior, boolean noSelec
...
-     * @param responder
      */
     protected final void doProcess(String referenceName, String mailboxName, ImapSession session, String tag, ImapCommand command, Responder responder, MailboxTyper mailboxTyper) {
         String user = ImapSessionUtils.getUserName(session);
-        final MailboxSession mailboxSession = ImapSessionUtils.getMailboxSession(session);
+        PathConverter pathConverter = PathConverter.forSession(session);
+        MailboxSession mailboxSession = ImapSessionUtils.getMailboxSession(session);
         try {
-            // Should the namespace section be returned or not?
-            final boolean isRelative;
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply protocols/imap/src/main/java/org/apache/james/imap/processor/ListProcessor.java]: The Cyrus model does not rely at all on relative handling.

To quote RFC-2342 : 

```
   IMAP4 [RFC-2060] does not define a default server namespace. As a
   result, two common namespace models have evolved:

   The "Personal Mailbox" model, in which the default namespace that is
   presented consists of only the user's personal mailboxes. To access
   shared mailboxes, the user must use an escape mechanism to reach
   another namespace.

   The "Complete Hierarchy" model, in which the default namespace that
   is presented includes the user's personal mailboxes along with any
   other mailboxes they have access to.
```

James was using "Complete Hierarchy" model. The goal of this ticket is to migrate it to "Complete Hierarchy model". Thus.... It makes no sens to continue  carring about relative handling.

[the remark this answers is not in the document]
Comment (mbaechler) [reply protocols/imap/src/main/java/org/apache/james/imap/processor/ListProcessor.java]: as said during meeting it deserves a specific commit explaining that.

**Comment by aduprat** [comment]
Build triggered. sha1 is original commit.

**Comment by aduprat** [comment]
Build triggered. sha1 is original commit.

**Comment by aduprat** [comment]
Build triggered. sha1 is original commit.

**Comment by openpaas-james** [comment]
Build triggered. sha1 is original commit.

**Comment by openpaas-james** [comment]
Build started sha1 is original commit.

**Comment by openpaas-james** [comment]
Build finished. No test results found.

**Comment by chibenwa** [comment]
Delegating support for IMAP is not a priority and won't be delivered any time soon

I'm closing this long outdated work.

**Pull request closed without merging**